### PR TITLE
Enable Redis connection pool if commons-pool2 is on the classpath

### PR DESCRIPTION
**Pull request by weixsun**
Is #26202  concise version

@snicoll 
I'm really sorry, I just want to make the commits look less, but it makes things more complicated.

There will be no changes before the merger or approval, please let things go back to normal, sorry again

### CODE REVIEW

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java [hunk 1]
```diff
@@ -95,6 +96,18 @@ private JedisClientConfigurationBuilder applyProperties(JedisClientConfiguration
...
+		if (poolEnabled) {
+			applyPooling(pool, builder);
+		}
+		final boolean isSentinelConfig = (getSentinelConfig() != null);
+		// Jedis Sentinel cannot operate without a pool.
+		if (!poolEnabled && isSentinelConfig) {
+			throw new RedisClientPoolingException(ClientType.JEDIS);
+		}
+	}
+
```
Comment (cuspymd) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java:110]: Wouldn't it be better to add `return;` after line 102 and remove the `!poolEnabled` check on line 106?

Comment (weixsun) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java]: I done it. Thank you for reviewing the code, you can see it in the latest commit

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java [hunk 2]
```diff
@@ -91,27 +95,38 @@ private LettuceConnectionFactory createLettuceConnectionFactory(LettuceClientCon
...
+	private LettuceClientConfigurationBuilder createBuilder(Environment environment, Pool pool) {
+		final boolean poolEnabled = environment.getProperty("spring.redis.lettuce.pool.enabled", Boolean.class, true);
+		if (poolEnabled) {
+			poolingFailureAnalyzerEvaluation();
+			return new PoolBuilderFactory().createBuilder(pool);
+		}
+		return LettuceClientConfiguration.builder();
+	}
+
+	private void poolingFailureAnalyzerEvaluation() {
```
Comment (cuspymd) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java:121]: Isn't it more clear that method names start with a verb? like`evaluatePoolingFailureAnalyzer()`

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingFailureAnalyzer.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			if (isGetDefaultPoolingFailureAction(message)) {
+				return getDefaultPoolingFailureAction(clientType);
+			}
+		}
+		return "";
+	}
+
+	private boolean isGetDefaultPoolingFailureAction(String message) {
+		return RedisClientPoolingException.JEDIS_SENTINEL_POOLING_EXPECTED_MESSAGE.equals(message)
+				|| RedisClientPoolingException.LETTUCE_LACK_COMMONSPOOL2_EXPECTED_MESSAGE.equals(message);
```
Comment (cuspymd) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingFailureAnalyzer.java:58]: It would be nice to give `RedisClientPoolingException` the responsibility of determining whether the exeption has a default message.

Comment (weixsun) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingFailureAnalyzer.java:58]: Here is the default corrective action when pooling fails.It is reasonable to put it in `RedisClientPoolingFailureAnalyzer`.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java (hunk 2 above)
Comment (weixsun) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java]: Indeed. Done it

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java [hunk 4]
```diff
@@ -95,6 +96,19 @@ private JedisClientConfigurationBuilder applyProperties(JedisClientConfiguration
 		return builder;
 	}
 
+	private void createPoolingBuilder(JedisClientConfigurationBuilder builder, Environment environment, Pool pool) {
+		final boolean poolEnabled = environment.getProperty("spring.redis.jedis.pool.enabled", Boolean.class, true);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java:100]: Given that `spring.redis.jedis.pool.enabled` is a property that you read in code, there's no reason to not map it in `Pool`. This will remove all those `Environment` injection points above.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java [hunk 5]
```diff
@@ -95,6 +96,19 @@ private JedisClientConfigurationBuilder applyProperties(JedisClientConfiguration
...
+		if (poolEnabled) {
+			applyPooling(pool, builder);
+			return;
+		}
+		final boolean isSentinelConfig = (getSentinelConfig() != null);
+		// Jedis Sentinel cannot operate without a pool.
+		if (isSentinelConfig) {
+			throw new RedisClientPoolingException(ClientType.JEDIS);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java:108]: That looks like something new to me and unrelated to configure the pool automatically. I am happy to consider this case at a later stage but for now please revert.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java [hunk 6]
```diff
@@ -73,12 +77,9 @@ private JedisConnectionFactory createJedisConnectionFactory(
...
 
 	private JedisClientConfiguration getJedisClientConfiguration(
-			ObjectProvider<JedisClientConfigurationBuilderCustomizer> builderCustomizers) {
+			ObjectProvider<JedisClientConfigurationBuilderCustomizer> builderCustomizers, Environment environment) {
 		JedisClientConfigurationBuilder builder = applyProperties(JedisClientConfiguration.builder());
-		RedisProperties.Pool pool = getProperties().getJedis().getPool();
-		if (pool != null) {
-			applyPooling(pool, builder);
-		}
+		createPoolingBuilder(builder, environment, getProperties().getJedis().getPool());
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java:82]: Let's try to put the changes to a minimum. There's no reason to create a new method. The pool != null in the caller can be changed by the check on the enabled flag and we should keep all the rest as it currently is.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java [hunk 7]
```diff
@@ -91,27 +95,38 @@ private LettuceConnectionFactory createLettuceConnectionFactory(LettuceClientCon
...
+			ClientResources clientResources, Pool pool, Environment environment) {
+		LettuceClientConfigurationBuilder builder = createBuilder(environment, pool);
+
 		applyProperties(builder);
 		if (StringUtils.hasText(getProperties().getUrl())) {
 			customizeConfigurationFromUrl(builder);
 		}
 		builder.clientOptions(createClientOptions());
 		builder.clientResources(clientResources);
+
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java:107]: Please remove.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java [hunk 8]
```diff
@@ -91,27 +95,38 @@ private LettuceConnectionFactory createLettuceConnectionFactory(LettuceClientCon
 
 	private LettuceClientConfiguration getLettuceClientConfiguration(
 			ObjectProvider<LettuceClientConfigurationBuilderCustomizer> builderCustomizers,
-			ClientResources clientResources, Pool pool) {
-		LettuceClientConfigurationBuilder builder = createBuilder(pool);
+			ClientResources clientResources, Pool pool, Environment environment) {
+		LettuceClientConfigurationBuilder builder = createBuilder(environment, pool);
+
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java:100]: Please remove.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java [hunk 9]
```diff
@@ -91,27 +95,38 @@ private LettuceConnectionFactory createLettuceConnectionFactory(LettuceClientCon
...
+		final boolean poolEnabled = environment.getProperty("spring.redis.lettuce.pool.enabled", Boolean.class, true);
+		if (poolEnabled) {
+			evaluatePoolingFailureAnalyzer();
+			return new PoolBuilderFactory().createBuilder(pool);
+		}
+		return LettuceClientConfiguration.builder();
+	}
+
+	private void evaluatePoolingFailureAnalyzer() {
+		final boolean pool2Present = ClassUtils.isPresent(PoolBuilderFactory.GENERIC_OBJECT_POOL_CONFIG_CLASS_NAME,
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java:122]: This should be a `private static final boolean` field

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java [hunk 10]
```diff
@@ -91,27 +95,38 @@ private LettuceConnectionFactory createLettuceConnectionFactory(LettuceClientCon
...
+				null);
+		if (!pool2Present) {
+			throw new RedisClientPoolingException(ClientType.LETTUCE);
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java:125]: There's no need for that. If the pool is not available, then no pool should be configured as before.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java [hunk 11]
```diff
@@ -91,27 +95,38 @@ private LettuceConnectionFactory createLettuceConnectionFactory(LettuceClientCon
...
 		}
-		return new PoolBuilderFactory().createBuilder(pool);
 	}
 
-	private LettuceClientConfigurationBuilder applyProperties(
-			LettuceClientConfiguration.LettuceClientConfigurationBuilder builder) {
+	private void applyProperties(LettuceClientConfiguration.LettuceClientConfigurationBuilder builder) {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/LettuceConnectionConfiguration.java:129]: Is there a reason to change the signature of this method? This is unrelated to this PR so I'd like this to be reverted.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingException.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.springframework.boot.autoconfigure.data.redis.RedisProperties.ClientType;
+import org.springframework.util.StringUtils;
+
+/**
+ * Exception thrown when a Redis client pooling failure.
+ *
+ * @author Weix Sun
+ */
+class RedisClientPoolingException extends RuntimeException {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingException.java:27]: Thanks but I don't think we need this for now.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingFailureAnalyzer.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+import org.springframework.boot.diagnostics.FailureAnalysis;
+import org.springframework.util.StringUtils;
+
+/**
+ * A {@code FailureAnalyzer} that performs analysis of failures caused by a
+ * {@link RedisClientPoolingException}.
+ *
+ * @author Weix Sun
+ */
+class RedisClientPoolingFailureAnalyzer extends AbstractFailureAnalyzer<RedisClientPoolingException> {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingFailureAnalyzer.java:30]: Same as the exception

File: spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 14]
```diff
@@ -1677,6 +1677,18 @@
         "replacement": "spring.reactor.debug-agent.enabled"
       }
     },
+    {
+      "name": "spring.redis.lettuce.pool.enabled",
+      "type": "java.lang.Boolean",
+      "description": "Whether to enable lettuce pool. If true, make sure \"commons-pool2\" exists on the classpath.",
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:1683]: I don't think there is a need to add the special hint there. There's no need to use `lettuce` in the description.

File: spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 15]
```diff
@@ -1677,6 +1677,18 @@
...
+      "defaultValue": true
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:1684]: There is no default value. It is enabled if `commons-pool2` is on the classpath.

File: spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 16]
```diff
@@ -1677,6 +1677,18 @@
         "replacement": "spring.reactor.debug-agent.enabled"
       }
     },
+    {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/additional-spring-configuration-metadata.json:1680]: This should move in code so that the metadata is generated automatically.

`Jedis#Pool` should have a `boolean` enabled field that's true by default.
`Lettuce#Pool` should have a `Boolean` enabled field that has no default value.

File: spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/spring.factories [hunk 17]
```diff
@@ -157,6 +157,7 @@ org.springframework.boot.autoconfigure.webservices.client.WebServiceTemplateAuto
 
 # Failure analyzers
 org.springframework.boot.diagnostics.FailureAnalyzer=\
+org.springframework.boot.autoconfigure.data.redis.RedisClientPoolingFailureAnalyzer,\
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/main/resources/META-INF/spring.factories:160]: Should be removed for the reasons above.

File: spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingFailureAnalyzerTests.java [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+import org.springframework.boot.diagnostics.FailureAnalysis;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Tests for {@link RedisClientPoolingFailureAnalyzer}.
+ *
+ * @author Weix Sun
+ */
+class RedisClientPoolingFailureAnalyzerTests {
```
Comment (snicoll) [inline spring-boot-project/spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingFailureAnalyzerTests.java:31]: Should be removed for now.

File: spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-data-redis/src/main/java/smoketest/data/redis/SampleRedisApplication.java [hunk 19]
```diff
@@ -40,8 +41,10 @@ public void run(String... args) throws Exception {
 	}
 
 	public static void main(String[] args) {
+		ConfigurableApplicationContext context = new SpringApplicationBuilder(SampleRedisApplication.class)
+				.properties("spring.redis.lettuce.pool.enabled:false").run(args);
```
Comment (snicoll) [inline spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-data-redis/src/main/java/smoketest/data/redis/SampleRedisApplication.java:45]: There's no reason to put the property here. Why not put it in `application.properties` as usual?

File: spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-data-redis/src/main/java/smoketest/data/redis/SampleRedisApplication.java [hunk 20]
```diff
@@ -40,8 +41,10 @@ public void run(String... args) throws Exception {
...
 		// Close the context so it doesn't stay awake listening for redis
-		SpringApplication.run(SampleRedisApplication.class, args).close();
+		context.close();
```
Comment (snicoll) [inline spring-boot-tests/spring-boot-smoke-tests/spring-boot-smoke-test-data-redis/src/main/java/smoketest/data/redis/SampleRedisApplication.java:47]: If the context does not properly shutdown when there is a pool, we may need to double check that we can't do anything about it.

**Review by snicoll — CHANGES_REQUESTED** [review]
Thanks for the PR @weixsun. I started to review in the aim of preparing this PR for our next feature release but there's too much work so I'd like you to have another look if possible. I suspect that the tests should be revisited as well.

Let me know if you have time. Thanks!

**Comment by weixsun** [comment]
@snicoll 
> there's too much work so I'd like you to have another look if possible

> Let me know if you have time. Thanks!

First of all, thank you for your review. You have indeed left a lot of suggestions for modification. Maybe it’s time to reach our goal again in an elegant and simple way. The good news is that I have plenty of time now, and I will reorganize and make appropriate modify.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java (hunk 5 above)
Comment (weixsun) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/JedisConnectionConfiguration.java]: I got it. Regarding this, I will report it in another issue or pr.

File: spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingException.java (hunk 12 above)
Comment (weixsun) [reply spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisClientPoolingException.java]: Got it.

**Comment by snicoll** [comment]
@weixsun those extra commits won't be reviewed as I've already reviewed the PR locally and polished things up. It is interesting though because I more or less did the same on my side.

**Review by snicoll — APPROVED** [review]

**Comment by weixsun** [comment]
@snicoll I got it. Experienced you can always be trusted. I'm sorry that I brought you extra work because of my original excessive thinking and coding.😅
